### ISSUE TITLE
Challenge #13 -Interactive visualization of HPC performance data

### ISSUE THREAD

**Issue by EsperanzaCuartero** [esowc/challenges_2020#3]
# Challenge #13 -  Interactive visualization of HPC performance data

> **Stream 1** - Weather related software and applications

> **IMPORTANT**: this challenge is eligible to apply for cloud credits from the European Weather Cloud.

### Goal
Develop interactive analysis tools for visualizing IFS performance data.

### Mentors and skills
* Mentors: @mlange05 , @marsdeno @crbrune
* **Skills required**
  * Python plotting libraries
  * Performance analysis (optional)

<hr>

### Challenge description
The continuous integration cycle of the IFS model is able to provide a regular stream of performance data, such as component runtimes, I/O and parallelisation overheads. This performance data is currently gathered and stored in a tabular format that can be converted to common data science formats, such as pandas.DataFrame. 

In this project we are aiming to develop a set of interactive visualization tools based on Python visualization packages (eg. matplotlib[1], Bokeh[2] or Altair[3]) to better track HPC performance of the model. An initial set of performance metrics, based on the IFS' internal performance monitoring tools (DrHook, GStats, ECProf) will be provided, which will be used to explore and demonstrate different interactive visualization interfaces. 

### References
* [1] [matplotlib plotting library](https://matplotlib.org/gallery/index.html) 
* [2] [Bokeh plotting library](https://docs.bokeh.org/en/latest/docs/gallery.html) 
* [3] [Altair plotting library](https://altair-viz.github.io/gallery/index.html)

**Comment by agiord** [comment]
@jwagemann, @mlange05  Hello, I'm very interested in this challenge! Is it possible to have some performance data sample to get to know them? Also, can you please provide some references in which hpc performance analysis of nwp models, possibly based on the metrics that will be provided, is carried out?
Thank you!

**Comment by tibi77** [comment]
Hello @mlange05 , @marsdeno!
From what I am gathering from the project description the project is supposed to create an interface that is going to respond to changes in the streams. Wouldn't it be something more interesting to make a tool that has the possibility to respond to immediate changes in the ifs? (You would also have the opportunity to look in the past.) I am asking because I have been working with a library for reactive programming in JavaScript and I am interested about the concept. Also I am interested about how DrHook works and how it all integrates in the IFS.

**Comment by mlange05** [comment]
@agiord and @tibi77 Thank you both for your interest in our ESoWC project! :smile: 

@agiord As a reference paper I would point to [this recent GMD paper](https://www.geosci-model-dev.net/12/4425/2019/) on the ESCAPE project. The paper includes some general plots (eg. Fig. 2, 6 and 8), but the project would not necessarily be limited to these plotting types. We are very interested in routinely exploring individual components of the model in depth and possibly compare and track those for different versions of the code. Some interactivity would also be desirable here, and, in fact, exploring different designs ideas for highlight certain aspects of model performance is the overall aim of this project.

Also, I will post some example data here later this week, so you can get a feel for the type of data sets.

@tibi77 Yes and no; while interactivity is a very desirable goal for a potential web-interface for exploring our performance data, the turnaround time for code merges and performance runs to create new data points is fairly large (several hours/days). So, the focus on interactivity was more in the context of interactive plots that act as a frontend to some stored performance data, where IFS developers can select metrics or components in certain plots to contrast different aspects of the model run. For example, one idea would be to allow IFS maintainers to "zoom in" from a high-level pie chart of model components to the individual subroutine calls that make up the aggregated time for that section. And again, the focus here is meant to be on exploring different plotting types to best express performance aspects.

As for DrHook, it's a very low-level, low-overhead C API that we inject in the model code to time and measure model performance at a variable granularity. It's output can be quite verbose and detailed, but we have utilities to filter it into `pandas.Dataframe` objects for downstream processing, which is what we would use as a starting point for this project.

**Comment by tibi77** [comment]
@mlange05 So an interface like the one that grafana[1] is offering would be a good starting point? Of course with more attention to the desired needs of the IFS developers.
Or the application should resemble more something like Intel Vtune what will point out the hot spots of the optimized code?  

Also, where is the data kept right now? Can I use this data in AWS or this is going to be an internal tool? 

[1] https://grafana.com/

**Comment by mlange05** [comment]
@tibi77 Yes, that is roughly the idea. I was personally envisaging a sort of "dashboard"-like webpage tailored to IFS with a bunch of Altair/Bokeh widgets, but yes the idea is roughly that. During initial development, a Jupyter notebook with some widgets for demonstration purposes would already be great, to highlight the different types of plots and animations that we would be interested in. How to go from widgets to a full dashboard is then a discussion for the later parts of the project.

The data itself is currently simply stored in a bunch of `.csv` files created from individual benchmark runs via `pandas.Dataframe`s. How we could organise the benchmark results in a more structured way would potentially also be an interesting discussion that we could look at once the project has progressed a little. And yes, for development purposes, we'll probably be using AWS for this project.

**Comment by mlange05** [comment]
[example_profile.zip](https://github.com/esowc/challenges_2020/files/4205454/example_profile.zip)

@agiord @tibi77 
Above is some example data to give you a flavour of what the processed output looks like (once unpacked this should be readable using `pandas.read_csv('example_profile.csv')`. It basically lists the min/max/accumulative runtime of individual subroutine routines across threads and MPI ranks. The subroutines are often nested, and the nesting information is not yet available in the file, but I'll add that before the project starts. We will hopefully also be able to provide some more detailed data (performance counters, etc.) once the project starts from other diagnostic tools.

**Comment by tibi77** [comment]
@mlange05 
Ok. Looked through the data and started making some initial scripts for diagnostics but I feel that they are hard to follow. I do not fully understand how the plots should be organised. I will make a github repo so I can experiment there with different plots. I am now thinking of something that will let you select the routine and see all the information from that routine in a simple plot. My problem is with the structure of that plot. How would you see it done? By the way is there information regarding the linking of the routines? (like which routine calls witch?) I am asking because I have been thinking of making a graph view for all of the routines and the when you press on one you can see the information regarding that routine.

**Comment by jwagemann** [comment]
Join us for our LIVE @ecmwf Summer of Weather Code Ask Me Anything session on 1 April 2020 at 2 pm (CET) (tomorrow).

Get infos first hand from the #ESoWC2020 organisers, mentors and former #ESoWC participants.
➡️[Sign up](https://esowc.ecmwf.int/ama)

**Comment by bs2603** [comment]
@jwagemann @mlange05 @marsdeno 
Hello!
I am interested in working on this project.
I went through the prerequisites and have prior experience with plotting high level usage as well as low level implementation of the plotting libraries.

In the csv file given by @mlange05 in a previous comment, can you please elucidate the kind of plots that are being expected?

**Comment by mlange05** [comment]
Hi @bs2603 and many thanks for you interest in this challenge. We are envisioning a dashboard-like set of widgets/interactive plots that can be used to analyse various performance aspects of the IFS model and its components. The design of the individual types and combinations of plots is hereby part of the challenge, as process will be driven by feedback and interaction with IFS developers and HPC performance engineers

To start start off with, I think some barcharts of individual routine runtimes and piecharts of accumulated runtimes for model components for comparison individual benchmark runs would be very useful, similar to some of the plots in this [GMD paper](https://www.geosci-model-dev.net/12/4425/2019/). Later we would probably add some roofline plots, and possibly some interactive re-scoping of piecharts, as well as scaling graphs and analyses of other low-level performance metrics. I realise that the data I posted earlier is not sufficient for some of this (it was only meant as a taster), as it doesn't capture some of the callgraph information needed for nested pie charts and has only very simple metrics, but we will hopefully post another example data set later this week which has some more information to play with.

**Comment by bs2603** [comment]
@mlange05  
[here](https://github.com/bs2603/ESoWC2020/tree/master/TestSamples) is a bar graph plot I've made 

Can you please take a look at this and tell what other features should I add for the given data. I've also uploaded the example.csv file I used.
I will make some more glyphs and upload them soon. 
The data you've provided gives the average run-times of the Routines,at the moment I've used that. I'll be eagerly waiting for data for individual run times and nested routines. 
I'd would also need the data to make pie charts as given in the GMD paper.

**Comment by jwagemann** [comment]
Only 4 days left to apply to be part of ECMWF Summer of Weather Code 2020. 
Application deadline: **Wednesday, 22 April 2020 at 23:59 (BST)**. 
Submit your proposal [here](https://esowc.commpla.com/form/application-form).

**Comment by mlange05** [comment]
@bs2603 Thanks, that first draft already looks very nice.

@bs2603 @tibi77 @agiord Here is another set of sample profiling data:
[more_profile_data.zip](https://github.com/esowc/challenges_2020/files/4503509/more_profile_data.zip)

This time the data contains individual routine metrics from two different runs (serial and thread-parallel) for a particular sub-component of the model. The data includes walltimes, as well as instruction metrics like IPC (instructions-per-cycle) and various PAPI counter readings for cache misses, etc. The zip file also contains a JSON file that maps individual routines to their individual call hierarchy, so so it should be possible to play around with visualising nested routine hierarchies. 

One final note, the toolchain that was used to create this data is in-house and fairly new, so performance metrics and raw output formats can all be adapted in the future. So, on top of exploring visualisation methods for this performance data, a co-design and feedback cycle for our internal development and profiling tools would be a valuable outcome of this project. :wink:
